### PR TITLE
Wallet format

### PR DESCRIPTION
**Pull request by AurelienFT**
* [ ] document all added functions
* [x] try in sandbox /simulation/labnet
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

Fix https://github.com/massalabs/massa/issues/3444

PR that make the same format of wallet in python in our deployment tooling : https://github.com/massalabs/private-deployment/pull/16 the wallet is well read by this PR

### CODE REVIEW

**Review by bilboquet — APPROVED** [review]
Only a comment about comments to fix :)

**Review by bilboquet — APPROVED** [review]
:+1:

File: massa-cipher/src/constants.rs [hunk 1]
```diff
@@ -6,17 +6,14 @@
...
 /// A nonce is a single-use value which enables securely encrypting multiple messages under the same key.
 /// Nonces need not be random: a counter can be used so long as the values are never repeated under the same key.
 pub const NONCE_SIZE: usize = 12;
 
 /// `PBKDF2` salt size.
-pub const SALT_SIZE: usize = 12;
+pub const SALT_SIZE: usize = 16;
 
 /// `PBKDF2` hash parameters.
 pub const HASH_PARAMS: Params = Params {
```
Comment (damip) [inline massa-cipher/src/constants.rs:19]: number of rounds does not follow spec: 600000

File: massa-cipher/src/encrypt.rs [hunk 2]
```diff
@@ -10,14 +10,19 @@ use pbkdf2::password_hash::Salt;
...
+}
 
 /// Encryption function using AES-GCM cipher.
 ///
 /// Read `lib.rs` module documentation for more information.
-pub fn encrypt(password: &str, data: &[u8]) -> Result<Vec<u8>, CipherError> {
+pub fn encrypt(password: &str, data: &[u8]) -> Result<CipherData, CipherError> {
     // generate the PBKDF2 salt
     let raw_salt: String = thread_rng()
         .sample_iter(&Alphanumeric)
```
Comment (damip) [inline massa-cipher/src/encrypt.rs:28]: does not follow spec: must be a sequence of bytes, not only alphanumeric which has way less entropy

Comment (damip) [reply massa-cipher/src/encrypt.rs:28]: note that the lib expects us to base64-encode the bytes to generate the salt

**Comment by damip** [comment]
To be tested:
* [ ] save a wallet with this code and try to load it with massa station
* [ ] save a wallet with massa station and try to load it with this code

**Review by damip — CHANGES_REQUESTED** [review]
Some critical security considerations from the spec are not applied

File: massa-cipher/src/decrypt.rs [hunk 3]
```diff
@@ -11,32 +11,17 @@ use pbkdf2::{
...
     // parse PBKDF2 salt
-    let salt_data = rest.get(..SALT_SIZE).ok_or_else(|| {
-        CipherError::DecryptionError(
-            "wallet file truncated: salt missing or incomplete".to_string(),
-        )
-    })?;
-    let salt = SaltString::new(std::str::from_utf8(salt_data)?)
+    let salt_data = data.salt;
+    let salt = SaltString::new(std::str::from_utf8(&salt_data)?)
         .map_err(|e| CipherError::DecryptionError(e.to_string()))?;
```
Comment (damip) [inline massa-cipher/src/decrypt.rs:25]: Note: I think we can use

```rust
let salt = SaltString::generate(&mut rng);
```

File: massa-cipher/src/encrypt.rs (hunk 2 above)
Comment (AurelienFT) [reply massa-cipher/src/encrypt.rs:28]: The problem is that the crate we use only accept &str and so utf8 chain of char which will not be the case if we use total random bytes

File: massa-cipher/src/constants.rs (hunk 1 above)
Comment (AurelienFT) [reply massa-cipher/src/constants.rs:19]: Done

**Comment by AurelienFT** [comment]
> To be tested before we merge:
> 
> * [ ]  save a wallet with this code and try to load it with massa station => compare keypairs
> * [ ]  save a wallet with massa station and try to load it with this code => compare keypairs

Will test when massa station will be up to date in testnet 24

File: massa-cipher/src/encrypt.rs (hunk 2 above)
Comment (damip) [reply massa-cipher/src/encrypt.rs:28]: I think that our crate expects us to provide `base64(random_bytes)`. Can you try and see if encryption on thyra + decryption in node works (and the opposite) works ?

**Comment by Eitu33** [comment]
@damip 
So after testing today with latest buildnet version, saving a massa-station wallet and importing it in the node massa-client produces this error:
```
Error: `MassaCipher` error: `Utf8` error: invalid utf-8 sequence of 1 bytes from index 0
```

It comes from `massa-cipher/src/decrypt.rs` at the salt parsing step:
```rust
let salt = SaltString::new(std::str::from_utf8(&salt_data)?)
    .map_err(|e| CipherError::DecryptionError(e.to_string()))?;
```
It expects an utf8 string and the massa-station wallet salt is composed of random bytes.

In the crate we use the salt implementation states the following:
```rust
/// Recommended length of a salt: 16-bytes.
///
/// This recommendation comes from the [PHC string format specification]:
///
/// > The role of salts is to achieve uniqueness. A *random* salt is fine
/// > for that as long as its length is sufficient; a 16-byte salt would
/// > work well (by definition, UUID are very good salts, and they encode
/// > over exactly 16 bytes). 16 bytes encode as 22 characters in B64.
```

Generating 16 bytes and encoding them as base64 solve two issues:
1. Fix the issue mentioned in https://github.com/massalabs/massa/pull/4044#discussion_r1226645490
2. Fix the incompatibility issue between massa-station and massa-cipher

But this requires us to make a change in the massa-station wallet encryption, adding a base64 encoding after the 16 random bytes generation.

We have to make a choice here:
1. Either keep what we have and add a base64 encoding step for the salt on both sides
2.

**Comment by gregLibert** [comment]
> Note: in the standard, we must not justify this requirement by "the rust lib needs it" but by the real reason behing it: to adhere to the [PHC string format specification](https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md#function-duties) recommended best practices.
> 
> More info here: https://docs.rs/password-hash/latest/password_hash/struct.Salt.html :
> 
> > The role of salts is to achieve uniqueness. A random salt is fine for that as long as its length is sufficient; a 16-byte salt would work well (by definition, UUID are very good salts, and they encode over exactly 16 bytes). 16 bytes encode as 22 characters in B64. Functions should disallow salt values that are too small for security (4 bytes should be viewed as an absolute minimum).

The standard we wrote is already following the good practice of `a 16-byte salt would work well`, see https://github.com/massalabs/massa-standards/blob/main/wallet/file-format.md#serialization.

**Comment by damip** [comment]
> > Note: in the standard, we must not justify this requirement by "the rust lib needs it" but by the real reason behing it: to adhere to the [PHC string format specification](https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md#function-duties) recommended best practices.
> > More info here: https://docs.rs/password-hash/latest/password_hash/struct.Salt.html :
> > > The role of salts is to achieve uniqueness. A random salt is fine for that as long as its length is sufficient; a 16-byte salt would work well (by definition, UUID are very good salts, and they encode over exactly 16 bytes). 16 bytes encode as 22 characters in B64. Functions should disallow salt values that are too small for security (4 bytes should be viewed as an absolute minimum).
> 
> The standard we wrote is already following the good practice of `a 16-byte salt would work well`, see https://github.com/massalabs/massa-standards/blob/main/wallet/file-format.md#serialization.

If Salt is a 16-byte vector of random bytes, the safety of `base64_encode(Salt)` is at least as good as that of `Salt` itself and can easily be supported in the Go wallet library. In the case of the rust library however, an extra layer of PHC-compliance is required and recommends the use base64-encoded salts.

We therefore have 3 options:
* implement different standards in Go vs Rust which will fragment the ecosystem
* update the standard to use base64, which makes it compatible with rust but needs a small update on the Go library to add the bs64 encoding
* fork the rust cryptographic library to remove the requirement for PHC compliance and maintain the fork

I like option 2 the most

**Comment by litchipi** [comment]
Ready to merge:
- This PR
- https://github.com/massalabs/station-massa-wallet/pull/596
- https://github.com/massalabs/massa-standards/pull/111
All at once, when this PR is reviewed and approved, and more tests are performed.

Test passing right now:
- Generating a wallet using the library in  `station-massa-wallet` (`wallet.Generate` then `wallet.Persist`), move it to `massa-clients/wallets`, start the client and run `wallet_info` to load the file

**Comment by litchipi** [comment]
@damip @bilboquet Ready for review before merging (with all the other PRs)

**Comment by litchipi** [comment]
Adapted the code to follow the new standard (unpadded base64)

>  Extra tests to make sure:
> 
>     load the wallet on both sides and ensure that the private AND public keys are the same
>     would be important to try the other way around as well: wallet generated by massa/massa-wallet and loaded by Station + perform the same checks above
> 

Tested these cases, it works well now :+1:

**Comment by damip** [comment]
If @AurelienFT confirms it's good to me :) congrats !

**Comment by litchipi** [comment]
This PR now gets the salt from the raw binary data `[u8; SALT_SIZE]` exported as `pub type Salt` from `massa-cipher`.
It converts it to the way we want before performing encryption / decryption operations.

No more changes are needed in the standard and in the massa-station side.
If you are OK with it @AurelienFT @damip , I'll close the other PRs once this PR is accepted and merged

**Comment by damip** [comment]
So now it works end-to-end between the node and massa-station without changes to the standard ?
If so, all good for me :)

**Comment by litchipi** [comment]
Yes it does

**Review by damip — APPROVED** [review]

**Pull request merged**